client/db.go

```diff
@@ -0,0 +1,461 @@
+package client
+
+import (
+	"encoding"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/testutils"
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+// Timestamp ...
+type Timestamp struct {
+	WallTime int64
+	Logical  int32
+}
+
+// KeyValue ...
+type KeyValue struct {
+	timestamp Timestamp
+	key       []byte
+	value     interface{}
+}
+
+// Key ...
+func (kv *KeyValue) Key(k interface{}) error {
+	switch t := k.(type) {
+	case encoding.BinaryUnmarshaler:
+		return t.UnmarshalBinary(kv.key)
+	case *[]byte:
+		*t = kv.key
+	case *string:
+		*t = string(kv.key)
+	}
+	return fmt.Errorf("unable to unmarshal value: %T", k)
+}
+
+// KeyBytes ...
+func (kv *KeyValue) KeyBytes() []byte {
+	return kv.key
+}
+
+// KeyString ...
+func (kv *KeyValue) KeyString() string {
+	return string(kv.key)
+}
+
+// Value ...
+func (kv *KeyValue) Value(v interface{}) error {
+	switch val := kv.value.(type) {
+	case []byte:
+		switch t := v.(type) {
+		case encoding.BinaryUnmarshaler:
+			return t.UnmarshalBinary(val)
+		case gogoproto.Message:
+			return gogoproto.Unmarshal(val, t)
+		case *[]byte:
+			*t = val
+		case *string:
+			*t = string(val)
+		}
+	case *int64:
+		switch t := v.(type) {
+		case *int64:
+			*t = *val
+		case *uint64:
+			*t = uint64(*val)
+		case *int32:
+			*t = int32(*val)
+		case *uint32:
+			*t = uint32(*val)
+		case *int:
+			*t = int(*val)
+		case *uint:
+			*t = uint(*val)
+		}
+	}
+	return fmt.Errorf("unable to unmarshal value: %T -> %T", kv.value, v)
+}
+
+func (kv *KeyValue) setValue(v *proto.Value) {
```

Comment:
There's no point in being allowed to set the timestamp on a value before putting it to the database. This will be disallowed if it doesn't match the timestamp of the actual PutRequest.